### PR TITLE
utilities.py refactor

### LINKED ISSUE #2838: Code complexity too high in utilities.py

### PR DESCRIPTION
**Pull request by MaxVanDeursen**
Fixes #2838

### CODE REVIEW

**Comment by devos50** [comment]
ok to test

**Comment by devos50** [comment]
@MaxVanDeursen please make sure the tests are succeeding and the code coverage/pylint checks are ok before putting your PR on ready. Also, please rebase your PR since there is a merge conflict with one of your changes 👍

**Comment by MaxVanDeursen** [comment]
@devos50 Ready for review

File: Tribler/Core/Utilities/utilities.py [hunk 1]
```diff
@@ -109,94 +179,164 @@ def validTorrentFile(metainfo):
...
+                raise ValueError('info files missing path or length key')
 
             p = file['path']
             if not isinstance(p, ListType):
                 raise ValueError('info files path is not list, but ' + repr(type(p)))
-            for dir in p:
-                if not isinstance(dir, StringType):
-                    raise ValueError('info files path is not string, but ' + repr(type(dir)))
+
+            if not all(isinstance(dirP, StringType) for dirP in p):
```
Comment (devos50) [inline Tribler/Core/Utilities/utilities.py]: Inconsistent variable name, please use underscore notation.

Comment (devos50) [inline Tribler/Core/Utilities/utilities.py]: Also, don't use one-letter variable names.

File: Tribler/Core/Utilities/utilities.py [hunk 2]
```diff
@@ -109,94 +179,164 @@ def validTorrentFile(metainfo):
...
+
+    The following information is validated:
+        -   Whether an info key exists within metainfo
+        -   Whether necessary keys exist within information (name, piece length and either root hash or pieces)
+        -   Whether the name value is of type String
+        -   Whether the piece length value is of type Long or Integer
+        -   Whether the root hash is an instance of String qnd has a length of 20 if it exists
+        -   Whether the pieces value is an instance of String and it has a length of multiple of 20
+
+    The validate_files function is used to validate more as well.
```
Comment (devos50) [inline Tribler/Core/Utilities/utilities.py]: This is not very descriptive. I don't think it's necessary to mention the `validate_files` method here.

File: Tribler/Core/Utilities/utilities.py [hunk 3]
```diff
@@ -109,94 +179,164 @@ def validTorrentFile(metainfo):
...
+    :raise ValueError: if one of the above validations do not succeed
+    """
+    if 'info' not in metainfo:
+        raise ValueError('metainfo misses key info')
+
+    info = metainfo['info']
+    if not isinstance(info, DictType):
+        raise ValueError('info not dict')
+
+    if not(all(x in info for x in {'name', 'piece length'}) and any(x in info for x in {'root hash', 'pieces'})):
```
Comment (devos50) [inline Tribler/Core/Utilities/utilities.py]: Please don't use one-letter variable names.

File: Tribler/Core/Utilities/utilities.py [hunk 4]
```diff
@@ -109,94 +179,164 @@ def validTorrentFile(metainfo):
...
+    pl = info['piece length']
+    if not (isinstance(pl, IntType) or isinstance(pl, LongType)):
+        raise ValueError('info piece size is not int, but ' + repr(type(pl)))
+
+    if 'root hash' in info:
+        rh = info['root hash']
+        if not isinstance(rh, StringType) or len(rh) != 20:
+            raise ValueError('info roothash is not 20-byte string')
+    else:
+        p = info['pieces']
```
Comment (devos50) [inline Tribler/Core/Utilities/utilities.py]: Please don't use one-letter variable names.

File: Tribler/Core/Utilities/utilities.py [hunk 5]
```diff
@@ -109,94 +179,164 @@ def validTorrentFile(metainfo):
...
+    :return: the original metainfo with corrected elements if possible
+    :raise ValueError: if there is a faulty element which cannot be corrected.
+    """
+    metainfo_result = metainfo
+
+    if not isinstance(metainfo, DictType):
+        raise ValueError('metainfo not dict')
+
+    # some .torrent files have a dht:// url in the announce field.
+    if 'announce' in metainfo and not (is_valid_url(metainfo['announce']) or metainfo['announce'].startswith('dht:')):
```
Comment (devos50) [inline Tribler/Core/Utilities/utilities.py]: Please use parenthesis to separate the and/or from each other and to make the order of evaluation more clear.

File: Tribler/Core/Utilities/utilities.py [hunk 6]
```diff
@@ -109,94 +179,164 @@ def validTorrentFile(metainfo):
...
-                    logger.warn("Warning: httpseeds url is not valid: %s HTTP seeding disabled", repr(url))
-                    break
+    metainfo_result['nodes'] = validate_torrent_nodes(metainfo_result)
+    metainfo_result['initial peers'] = validate_init_peers(metainfo)
+    metainfo_result['url-list'] = validate_url_list(metainfo)
+    metainfo_result['httpseeds'] = validate_http_seeds(metainfo)
+    metainfo_result['info'] = validate_torrent_info(metainfo)
+
+    # remove elements if None i.e. not valid.
+    for x in {'httpseeds', 'url-list', 'nodes', 'initial peers'}:
```
Comment (devos50) [inline Tribler/Core/Utilities/utilities.py]: Please don't use one-letter variable names.

File: Tribler/Core/Utilities/utilities.py [hunk 7]
```diff
@@ -109,94 +179,164 @@ def validTorrentFile(metainfo):
...
+        if not metainfo[x]:
+            del metainfo[x]
+
+    if not ('announce' in metainfo or 'nodes' in metainfo):
+        # disabling this check, modifying metainfo to allow for ill-formatted torrents
+        metainfo_result['nodes'] = []
+
+    return dict((k, v) for k, v in metainfo_result.iteritems() if v or (metainfo[k] and metainfo[k] == v))
```
Comment (devos50) [inline Tribler/Core/Utilities/utilities.py]: Please don't use one-letter variable names.

File: Tribler/Core/Utilities/utilities.py [hunk 8]
```diff
@@ -109,94 +179,164 @@ def validTorrentFile(metainfo):
...
-def isValidTorrentFile(metainfo):
+    :param metainfo: the metainfo to be validated
+    :return: Boolean specifying whether the specified metainfo is valid or not
+    """
     try:
-        validTorrentFile(metainfo)
+        create_valid_metainfo(metainfo)
         return True
     except ValueError:
         logger.exception("Could not check torrent file: a ValueError was thrown")
```
Comment (devos50) [inline Tribler/Core/Utilities/utilities.py]: Please modify this log message (something like "Torrent file is invalid").

File: Tribler/Core/Utilities/utilities.py [hunk 9]
```diff
@@ -109,94 +179,164 @@ def validTorrentFile(metainfo):
...
+    :return: Boolean specifying whether the URL is valid
+    """
+    if url.lower().startswith('udp'):
         url = url.lower().replace('udp', 'http', 1)
     r = urlparse.urlsplit(url)
 
-    if r[0] == '' or r[1] == '':
-        return False
-    return True
+    return not(r[0] == '' or r[1] == '')
```
Comment (devos50) [inline Tribler/Core/Utilities/utilities.py]: Please don't use one-letter variable names.

File: Tribler/Core/Utilities/utilities.py [hunk 10]
```diff
@@ -109,94 +179,164 @@ def validTorrentFile(metainfo):
...
 
 
 def http_get(uri):
-
+    """
+    Performs a GET request
+    :param uri: The URL to perform a GET request to
+    :return: A deferred returning the body of the response.
```
Comment (devos50) [inline Tribler/Core/Utilities/utilities.py]: The deferred does not return the body of the response; it *fires* with the body of the response.

File: Tribler/Core/Utilities/utilities.py [hunk 11]
```diff
@@ -109,94 +179,164 @@ def validTorrentFile(metainfo):
...
+    :raises HttpError: When the HTTP response code is not OK (i.e. 200)
```
Comment (devos50) [inline Tribler/Core/Utilities/utilities.py]: 200 means ok?

File: Tribler/Test/Core/test_torrent_def.py [hunk 12]
```diff
@@ -26,6 +26,7 @@
 class TestTorrentDef(BaseTestCase):
 
     VIDEO_FILE_NAME = "video.avi"
+    maxDiff = None
```
Comment (devos50) [inline Tribler/Test/Core/test_torrent_def.py]: Please use underscore notation for variable names.

Comment (devos50) [inline Tribler/Test/Core/test_torrent_def.py]: Also, what's the purpose of this variable? It appears to be unused.

File: Tribler/Core/Utilities/utilities.py [hunk 13]
```diff
@@ -1,7 +1,10 @@
+"""
+This class mainly provides validation and correction for both metainfo and urls. This class
```
Comment (MaxVanDeursen) [inline Tribler/Core/Utilities/utilities.py]: This is [PEP 257](http://legacy.python.org/dev/peps/pep-0257/#documentation-strings):

> The docstring for a module should generally list the classes, exceptions and functions (and any other objects) that are exported by the module, with a one-line summary of each. (These summaries generally give less detail than the summary line in the object's docstring.)

File: Tribler/Core/Utilities/utilities.py [hunk 14]
```diff
@@ -75,35 +79,93 @@ def validTorrentFile(metainfo):
...
-        if not isinstance(rh, StringType) or len(rh) != 20:
-            raise ValueError('info roothash is not 20-byte string')
-    else:
-        p = info['pieces']
-        if not isinstance(p, StringType) or len(p) % 20 != 0:
-            raise ValueError('info pieces is not multiple of 20 bytes')
+    First checks whether a URL-list exists, after which it is checked whether the metainfo
+    would be able to use HTTP seeding. This is not the case if there are multiple files specified.
+
+    A warning is logged if:
```
Comment (MaxVanDeursen) [inline Tribler/Core/Utilities/utilities.py]: Although this is specified in the `return` line of the docstring, I've made sure to make that more clear.

File: Tribler/Core/Utilities/utilities.py [hunk 15]
```diff
@@ -21,25 +20,20 @@
...
-    # if 'announce' in metainfo and 'nodes' in metainfo:
-    #    raise ValueError('both announce and nodes present')
+    First it is checked whether metainfo contains a list of nodes.
+    After this, for each node in the list, the following is checked:
+        - Whether the node is a pair of length two
+        - Whether the first element is of type String
+        - Whether the second element is a tuple consisting of an Int and a Long
 
+    :param metainfo: the metainfo for which the nodes have to be validated
+    :return: Returns the nodes of the metainfo or None if they don't exist
```
Comment (MaxVanDeursen) [inline Tribler/Core/Utilities/utilities.py]: Would you have a suggestion as to what to use as a general name for these methods? I could not think of anything better than validate. Moreover, one could read the docstring to know what the function returns.

File: Tribler/Test/Core/test_utilities.py [hunk 16]
```diff
@@ -1,220 +1,220 @@
...
+from Tribler.Core.Utilities.utilities import create_valid_metainfo, valid_torrent_file, parse_magnetlink
 from Tribler.Test.Core.base_test import TriblerCoreTest
 
 
 class TriblerCoreTestUtilities(TriblerCoreTest):
 
     @raises(ValueError)
     def test_valid_torrent_file_invalid_metainfo_type(self):
-        validTorrentFile([])
+        create_valid_metainfo([])
```
Comment (MaxVanDeursen) [inline Tribler/Test/Core/test_utilities.py]: Done that

**Comment by devos50** [comment]
@MaxVanDeursen please rebase this PR and I will merge it.

**Comment by MaxVanDeursen** [comment]
Changes rebased

**Comment by devos50** [comment]
Thanks! 👍

**Pull request merged**